rust/arrow/src/compute/kernels/take.rs

```diff
@@ -0,0 +1,519 @@
+        DataType::Timestamp(Millisecond) => {
+            take_numeric::<TimestampMillisecondType>(array, index)
+        }
+        DataType::Timestamp(Microsecond) => {
+            take_numeric::<TimestampMicrosecondType>(array, index)
+        }
+        DataType::Timestamp(Nanosecond) => {
+            take_numeric::<TimestampNanosecondType>(array, index)
+        }
+        DataType::Utf8 => take_binary(array, index),
+        DataType::List(_) => take_list(array, index),
+        DataType::Struct(fields) => {
+            let struct_: &StructArray =
+                array.as_any().downcast_ref::<StructArray>().unwrap();
+            let arrays: Result<Vec<ArrayRef>> = struct_
+                .columns()
+                .iter()
+                .map(|a| take(a, index, Some(&options)))
+                .collect();
+            let arrays = arrays?;
+            let pairs: Vec<(Field, ArrayRef)> =
+                fields.clone().into_iter().zip(arrays).collect();
+            Ok(Arc::new(StructArray::from(pairs)) as ArrayRef)
+        }
+        t @ _ => unimplemented!("Sort not supported for data type {:?}", t),
+    }
+}
+
+/// Options that define how `take` should behave
+#[derive(Clone)]
+pub struct TakeOptions {
+    /// perform bounds check before taking
+    pub check_bounds: bool,
+}
+
+impl Default for TakeOptions {
+    fn default() -> Self {
+        Self {
+            check_bounds: false,
+        }
+    }
+}
+
+/// `take` implementation for numeric arrays
+fn take_numeric<T>(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef>
+where
+    T: ArrowNumericType,
+{
+    let mut builder = PrimitiveBuilder::<T>::new(index.len());
+    let a = array.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append_null()?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_valid(ix) {
+                builder.append_value(a.value(ix))?;
+            } else {
+                builder.append_null()?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for binary arrays
+fn take_binary(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
+    let mut builder = BinaryBuilder::new(index.len());
+    let a = array.as_any().downcast_ref::<BinaryArray>().unwrap();
+    for i in 0..index.len() {
+        if index.is_null(i) {
+            builder.append(false)?;
+        } else {
+            let ix = index.value(i) as usize;
+            if a.is_null(ix) {
+                builder.append(false)?;
+            } else {
+                builder.append_values(a.value(ix))?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for boolean arrays
+fn take_bool(array: &ArrayRef, index: &UInt32Array) -> Result<ArrayRef> {
```

Comment:
Can this be merged with `take_numeric`?

Reply:
`fn value(&self, i: usize)` is implemented separately for `ArrowNumericType` bound and `BooleanType`, so I (think) I can't create a common `take_primitive`. If I say `... where T: ArrowPrimitiveType`, I fall short because I need the `fn value()`

Reply:
I see. I encountered the exact same issue in another PR. The approach I took is to implement an internal trait just for the `value()` method and have both `PrimitiveArray` and `BooleanArray` implementing it. You can either do it here or we can improve this in a follow-up.

Reply:
I'm looking at it now, I can address it in a follow-up as there are other places that can benefit from this dedup.

Reply:
I ended up using the internal trait that you created, and have merged numeric with boolean into primitive.